File: GoogleMaps/GoogleMapsXCFrameworkDemos/Samples/DataDrivenStylingSearchViewController.m
/*
 * Copyright 2023 Google LLC. All rights reserved.
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
 * file except in compliance with the License. You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
 * ANY KIND, either express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

#import "GoogleMapsXCFrameworkDemos/Samples/DataDrivenStylingSearchViewController.h"

#import <Foundation/Foundation.h>
#import <UIKit/UIKit.h>

#import "GoogleMapsXCFrameworkDemos/Common/GMSNotCapturingTouchesTableView.h"
#import "GoogleMapsXCFrameworkDemos/Common/UIViewController+GMSModals.h"
#import "GoogleMapsXCFrameworkDemos/SDKDemoAPIKey.h"
#if __has_feature(modules)
@import GoogleMaps;
#else
#import <GoogleMaps/GoogleMaps.h>
#endif

NS_ASSUME_NONNULL_BEGIN

// Put your Map ID in the string below. In the cloud console, configure the Map ID with a style that
// enables the "Administrative Area Level 2" feature layer.
static NSString *const kMapIDWithAdmin2 = @"";

static NSArray<NSString *> *GetInitialSearches(void) {
  return @[ @"Nye County", @"San Bernardino County", @"Juab County", @"Crook County" ];
}

// This demo uses Text Search feature from Places API. The API must be enabled from Cloud Console.
// See https://developers.google.com/maps/documentation/places/web-service/search-textual for
// details.
static NSURLRequest *BuildSearchRequestForPlaceName(NSString *placeName) {
  // NSURL initializer only returns nil when the URL string is malformed, but it is a known valid
  // string literal here.
  NSURL *URL =
      (NSURL *_Nonnull)[NSURL URLWithString:@"https://places.googleapis.com/v1/places:searchText"];
  NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:URL];
  request.HTTPMethod = @"POST";
  [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
  [request setValue:kAPIKey forHTTPHeaderField:@"X-Goog-Api-Key"];
  [request setValue:@"places.id" forHTTPHeaderField:@"X-Goog-FieldMask"];
  [request setValue:[[NSBundle mainBundle] bundleIdentifier]
      forHTTPHeaderField:@"X-Ios-Bundle-Identifier"];
  NSDictionary<NSString *, NSString *> *requestBody = @{
    @"textQuery" : placeName,
    @"includedType" : @"administrative_area_level_2",
    @"languageCode" : @"en"
  };
  request.HTTPBody = [NSJSONSerialization dataWithJSONObject:requestBody
                                                     options:(NSJSONWritingOptions)0
                                                       error:nil];
  return request;
}

static NSString *_Nullable ExtractPlaceIDFromSearchResponse(id JSONResponse) {
  if (![JSONResponse isKindOfClass:[NSDictionary class]]) {
    return nil;
  }
  id placesArray = ((NSDictionary *)JSONResponse)[@"places"];
  if (![placesArray isKindOfClass:[NSArray class]]) {
    return nil;
  }
  id place = placesArray[0];
  if (![place isKindOfClass:[NSDictionary class]]) {
    return nil;
  }
  NSString *ID = place[@"id"];
  if (![ID isKindOfClass:[NSString class]]) {
    return nil;
  }
  return ID;
}

@interface GMSPlaceLookupController
    : NSObject <UITextFieldDelegate, UIColorPickerViewControllerDelegate>

@property(nonatomic, readonly) NSUInteger serial;
@property(nonatomic, nullable) NSString *name;
@property(nonatomic, nullable) UIColor *color;
@property(nonatomic, readonly, nullable) NSString *placeID;

@property(nonatomic, readonly) UIView *view;

- (void)focusTextEdit;

@end

@interface DataDrivenStylingSearchViewController ()

@property(nonatomic, readonly) NSURLSession *URLSession;

- (void)removeFeature:(GMSPlaceLookupController *)feature;

- (void)reloadStyle;

@end

@implementation GMSPlaceLookupController {
  __weak DataDrivenStylingSearchViewController *_controller;

  UITextField *_textField;
  UIButton *_colorSelectionButton;
}

- (instancetype)initWithController:(DataDrivenStylingSearchViewController *)controller
                            serial:(NSUInteger)serial {
  self = [super init];
  if (self) {
    _controller = controller;
    _serial = serial;

    _textField = [[UITextField alloc] init];
    _textField.translatesAutoresizingMaskIntoConstraints = NO;
    _textField.backgroundColor = [UIColor.whiteColor colorWithAlphaComponent:0.5];
    _textField.font = [_textField.font fontWithSize:12];
    _textField.delegate = self;

    _colorSelectionButton = [[UIButton alloc] initWithFrame:CGRectZero];
    _colorSelectionButton.titleLabel.font = [_colorSelectionButton.titleLabel.font fontWithSize:12];
    [_colorSelectionButton setTitle:@"❓" forState:UIControlStateNormal];
    [_colorSelectionButton addTarget:self
                              action:@selector(selectionButtonTapped:)
                    forControlEvents:UIControlEventTouchUpInside];

    _textField.leftView = _colorSelectionButton;
    _textField.leftViewMode = UITextFieldViewModeAlways;

    UIButton *clearButton = [[UIButton alloc] initWithFrame:CGRectZero];
    clearButton.titleLabel.font = [clearButton.titleLabel.font fontWithSize:12];
    [clearButton setTitle:@"🗑️" forState:UIControlStateNormal];
    [clearButton addTarget:self
                    action:@selector(remove)
          forControlEvents:UIControlEventTouchUpInside];
    _textField.rightView = clearButton;
    _textField.rightViewMode = UITextFieldViewModeUnlessEditing;

    _textField.clearButtonMode = UITextFieldViewModeWhileEditing;
  }
  return self;
}

- (UIView *)view {
  return _textField;
}

- (void)setPlaceID:(NSString *)placeID {
  _placeID = [placeID copy];
  [_colorSelectionButton setTitle:@"⬤" forState:UIControlStateNormal];
  [_controller reloadStyle];
}

- (void)setName:(nullable NSString *)name {
  _name = [name copy];
  _textField.text = name;

  if (!name || name.length == 0) {
    return;
  }

  __weak __typeof__(self) weakSelf = self;
  NSString *nonNilName = name;
  NSURLSessionDataTask *dataTask = [_controller.URLSession
      dataTaskWithRequest:BuildSearchRequestForPlaceName(nonNilName)
        completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
          __weak __typeof__(self) strongSelf = weakSelf;
          if (!strongSelf) {
            return;
          }
          if (error != nil) {
            [strongSelf textSearchFailedWithError:error];
            return;
          }
          NSError *JSONError = nil;
          id deserialized = [NSJSONSerialization JSONObjectWithData:data
                                                            options:(NSJSONReadingOptions)0
                                                              error:&JSONError];
          if (JSONError != nil) {
            [strongSelf textSearchFailedWithError:JSONError];
            return;
          }
          NSString *ID = ExtractPlaceIDFromSearchResponse(deserialized);
          if (ID) {
            strongSelf.placeID = ID;
          } else {
            NSError *unrecognizedResponseError =
                [NSError errorWithDomain:NSStringFromClass([strongSelf class])
                                    code:-1
                                userInfo:@{@"response" : deserialized}];
            [strongSelf textSearchFailedWithError:unrecognizedResponseError];
          }
        }];
  [dataTask resume];
}

- (void)setColor:(nullable UIColor *)color {
  _color = color;
  [_colorSelectionButton setTitleColor:color forState:UIControlStateNormal];
  [_controller reloadStyle];
}

- (void)focusTextEdit {
  [_textField becomeFirstResponder];
}

- (void)remove {
  [_controller removeFeature:self];
}

- (void)textSearchFailedWithError:(NSError *)error {
  [_colorSelectionButton setTitle:@"❗" forState:UIControlStateNormal];
  NSLog(@"Feature \"%@\" not found: %@", self.name, error);
}

- (void)selectionButtonTapped:(UIButton *)sender {
  UIColorPickerViewController *selector = [[UIColorPickerViewController alloc] init];
  selector.selectedColor = _color ?: UIColor.whiteColor;
  selector.supportsAlpha = NO;
  selector.delegate = self;
  [_controller presentModalViewController:selector animated:YES];
}

#pragma mark - UITextFieldDelegate

- (void)textFieldDidEndEditing:(UITextField *)textField {
  NSString *text = [textField.text
      stringByTrimmingCharactersInSet:NSCharacterSet.whitespaceAndNewlineCharacterSet];
  if (!text.length) {
    [_colorSelectionButton setTitle:@"❓" forState:UIControlStateNormal];
    return;
  }
  if ([self.name isEqualToString:text]) {
    return;
  }
  [_colorSelectionButton setTitle:@"⏳" forState:UIControlStateNormal];
  self.name = text;
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
  // Finishes editing when pressing Enter.
  [textField resignFirstResponder];
  return NO;
}

#pragma mark - UIColorPickerViewControllerDelegate

- (void)colorPickerViewControllerDidFinish:(UIColorPickerViewController *)viewController {
  self.color = viewController.selectedColor;
}

@end

@implementation DataDrivenStylingSearchViewController {
  GMSMapView *_mapView;
  UITableViewDiffableDataSource *_dataSource;
  UITableView *_tableView;

  NSUInteger _placeSerial;

  NSMutableDictionary<NSNumber *, GMSPlaceLookupController *> *_places;
  NSDictionary<NSString *, UIColor *> *_computedColorMapping;
}

- (instancetype)init {
  self = [super init];
  if (self) {
    _places = [NSMutableDictionary dictionary];
    _URLSession = [NSURLSession
        sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]
                        delegate:nil
                   delegateQueue:[NSOperationQueue mainQueue]];
  }
  return self;
}

- (void)viewDidLoad {
  [super viewDidLoad];

  if (kMapIDWithAdmin2.length) {
    [self setUpMapWithMapID:kMapIDWithAdmin2];
    return;
  }

  __weak __typeof__(self) weakSelf = self;
  [self gms_promptForMapIDWithDescription:@"with Administrative Area Level 2 layer enabled"
                                  handler:^(NSString *mapID) {
                                    [weakSelf setUpMapWithMapID:mapID];
                                  }];
}

- (void)setUpMapWithMapID:(NSString *)mapIDString {
  if (!mapIDString.length) {
    UILabel *text = [[UILabel alloc] initWithFrame:CGRectZero];
    text.text = @"A Map ID is required";
    text.textAlignment = NSTextAlignmentCenter;
    self.view = text;
    return;
  }

  GMSCameraPosition *camera = [GMSCameraPosition cameraWithLatitude:40 longitude:-117.5 zoom:5.5];
  _mapView = [GMSMapView mapWithFrame:CGRectZero
                                mapID:[GMSMapID mapIDWithIdentifier:mapIDString]
                               camera:camera];
  [self.view addSubview:_mapView];
  _mapView.translatesAutoresizingMaskIntoConstraints = NO;
  [NSLayoutConstraint activateConstraints:@[
    [_mapView.topAnchor constraintEqualToAnchor:self.view.topAnchor],
    [_mapView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
    [_mapView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
    [_mapView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor],
  ]];

  NSString *cellIdentifier = NSStringFromClass([self class]);
  _tableView = [[GMSNotCapturingTouchesTableView alloc] initWithFrame:CGRectZero
                                                                style:UITableViewStylePlain];
  [_tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:cellIdentifier];
  _tableView.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentNever;
  _tableView.rowHeight = 28;
  _tableView.backgroundView = nil;
  _tableView.backgroundColor = UIColor.clearColor;

  NSMutableDictionary<NSNumber *, GMSPlaceLookupController *> *places = _places;
  _dataSource = [[UITableViewDiffableDataSource alloc]
      initWithTableView:_tableView
           cellProvider:^UITableViewCell *(UITableView *tableView, NSIndexPath *indexPath,
                                           NSNumber *itemIdentifier) {
             UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier
                                                                     forIndexPath:indexPath];
             [cell.contentView.subviews makeObjectsPerformSelector:@selector(removeFromSuperview)];
             UIView *view = places[itemIdentifier].view;
             [cell.contentView addSubview:view];
             cell.backgroundColor = [UIColor clearColor];
             [NSLayoutConstraint activateConstraints:@[
               [cell.contentView.topAnchor constraintEqualToAnchor:view.topAnchor],
               [cell.contentView.leadingAnchor constraintEqualToAnchor:view.leadingAnchor],
               [cell.contentView.trailingAnchor constraintEqualToAnchor:view.trailingAnchor],
               [cell.contentView.bottomAnchor constraintEqualToAnchor:view.bottomAnchor]
             ]];

             return cell;
           }];
  _tableView.dataSource = _dataSource;
  _tableView.translatesAutoresizingMaskIntoConstraints = NO;
  [self.view addSubview:_tableView];

  __weak __typeof__(self) weakSelf = self;
  UIAction *addAction = [UIAction actionWithTitle:@"+"
                                            image:nil
                                       identifier:nil
                                          handler:^(UIAction *action) {
                                            [weakSelf addButtonTapped];
                                          }];
  self.navigationItem.rightBarButtonItem =
      [[UIBarButtonItem alloc] initWithPrimaryAction:addAction];

  [NSLayoutConstraint activateConstraints:@[
    [_tableView.topAnchor constraintEqualToAnchor:self.view.topAnchor],
    [_tableView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
    [_tableView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
    [_tableView.heightAnchor constraintEqualToAnchor:self.view.heightAnchor multiplier:0.5]
  ]];

  NSArray<NSString *> *initialSearches = GetInitialSearches();
  NSMutableArray<NSNumber *> *identifiers =
      [NSMutableArray arrayWithCapacity:initialSearches.count];
  for (NSString *search in initialSearches) {
    [identifiers addObject:@([self addFeatureWithText:search])];
  }
  NSDiffableDataSourceSnapshot *diff = [[NSDiffableDataSourceSnapshot alloc] init];
  [diff appendSectionsWithIdentifiers:@[ @0 ]];
  [diff appendItemsWithIdentifiers:identifiers];
  [_dataSource applySnapshot:diff animatingDifferences:NO];
}

- (void)addButtonTapped {
  NSNumber *itemIdentifier = @([self addFeatureWithText:nil]);
  GMSPlaceLookupController *lookup = _places[itemIdentifier];

  NSDiffableDataSourceSnapshot *diff = [_dataSource snapshot];
  [diff appendItemsWithIdentifiers:@[ itemIdentifier ]];
  [_dataSource applySnapshot:diff animatingDifferences:NO];

  dispatch_async(dispatch_get_main_queue(), ^{
    [lookup focusTextEdit];
  });
}

- (CGFloat)nextColorHue {
  NSMutableArray<NSNumber *> *hueValues = [NSMutableArray arrayWithCapacity:_places.count];
  for (GMSPlaceLookupController *config in _places.objectEnumerator) {
    CGFloat hueValue = 0.f;
    if ([config.color getHue:&hueValue saturation:nil brightness:nil alpha:nil]) {
      if (hueValue > 1 || hueValue < 0) {
        continue;
      }
      [hueValues addObject:@(hueValue)];
    }
  }
  float candidateHue = hueValues.firstObject.floatValue - 0.5f;
  if (hueValues.count > 1) {
    [hueValues sortUsingSelector:@selector(compare:)];
    CGFloat endOfLargestGap = hueValues.firstObject.floatValue;
    CGFloat largestGap = 1 + endOfLargestGap - hueValues.lastObject.floatValue;
    for (NSUInteger i = 1; i < hueValues.count; i++) {
      CGFloat gap = hueValues[i].floatValue - hueValues[i - 1].floatValue;
      if (gap > largestGap) {
        endOfLargestGap = hueValues[i].floatValue;
        largestGap = gap;
      }
    }
    candidateHue = endOfLargestGap - largestGap / 2;
  }
  if (candidateHue < 0) {
    candidateHue += 1.f;
  }
  return candidateHue;
}

- (NSUInteger)addFeatureWithText:(nullable NSString *)text {
  NSUInteger itemIdentifier = ++_placeSerial;

  GMSPlaceLookupController *featureConfig =
      [[GMSPlaceLookupController alloc] initWithController:self serial:itemIdentifier];

  featureConfig.color = [UIColor colorWithHue:[self nextColorHue]
                                   saturation:1
                                   brightness:0.75
                                        alpha:1];
  featureConfig.name = text;
  _places[@(itemIdentifier)] = featureConfig;

  return itemIdentifier;
}

- (void)removeFeature:(GMSPlaceLookupController *)feature {
  if (_places.count == 1) {
    return;
  }

  NSDiffableDataSourceSnapshot *diff = [_dataSource snapshot];
  [diff deleteItemsWithIdentifiers:@[ @(feature.serial) ]];
  [_dataSource applySnapshot:diff animatingDifferences:NO];

  [_places removeObjectForKey:@(feature.serial)];
  [self reloadStyle];
}

- (void)reloadStyle {
  NSMutableDictionary<NSString *, UIColor *> *colorMapping = [NSMutableDictionary dictionary];
  for (GMSPlaceLookupController *config in _places.objectEnumerator) {
    NSString *placeID = config.placeID;
    if (placeID != nil) {
      colorMapping[placeID] = config.color;
    }
  }
  if ([_computedColorMapping isEqualToDictionary:colorMapping]) {
    // Skip restyling if the mapping hasn't actually changed, since restyling can be expensive.
    return;
  }
  _computedColorMapping = colorMapping;

  [_mapView featureLayerOfFeatureType:GMSFeatureTypeAdministrativeAreaLevel2].style =
      ^GMSFeatureStyle *(GMSPlaceFeature *feature) {
    UIColor *color = colorMapping[feature.placeID];
    if (!color) {
      return nil;
    }
    return [GMSFeatureStyle styleWithFillColor:[color colorWithAlphaComponent:0.5]
                                   strokeColor:color
                                   strokeWidth:1.5];
  };
}

@end

NS_ASSUME_NONNULL_END

File: GoogleMaps-Swift/GoogleMapsSwiftXCFrameworkDemos/Swift/Samples/DataDrivenStylingSearchViewController.swift
// Copyright 2022 Google LLC. All rights reserved.
//
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
// file except in compliance with the License. You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under
// the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
// ANY KIND, either express or implied. See the License for the specific language governing
// permissions and limitations under the License.

import GoogleMaps
import UIKit

/// Put your Map ID in the string below. In the cloud console, configure the Map ID with a style
/// that enables the "Administrative Area Level 2" feature layer.
private let mapIDWithAdministrativeAreaLevel2Layer = ""

private let initialSearches = [
  "Nye County", "San Bernardino County", "Juab County", "Crook County",
]

private func buildSearchRequest(forPlaceName placeName: String) -> URLRequest {
  // URL initializer only returns nil when the URL string is malformed, but it is a known valid
  // string literal here.
  let requestURL = URL(string: "https://places.googleapis.com/v1/places:searchText")!
  var urlRequest = URLRequest(url: requestURL)
  urlRequest.httpMethod = "POST"
  urlRequest.addValue("application/json", forHTTPHeaderField: "Content-Type")
  urlRequest.addValue(SDKConstants.apiKey, forHTTPHeaderField: "X-Goog-Api-Key")
  urlRequest.addValue("places.id", forHTTPHeaderField: "X-Goog-FieldMask")
  urlRequest.addValue(
    Bundle.main.bundleIdentifier ?? "unknown_ios", forHTTPHeaderField: "X-Ios-Bundle-Identifier")
  urlRequest.httpBody = try! JSONSerialization.data(
    withJSONObject: [
      "textQuery": placeName,
      "includedType": "administrative_area_level_2",
      "languageCode": "en",
    ], options: [])
  return urlRequest
}

private struct PlaceTextSearchResponse: Codable {
  struct Place: Codable {
    let id: String
  }

  let places: [Place]
}

private class PlaceLookupController: NSObject {
  let serial: Int
  let controller: DataDrivenStylingSearchViewController

  var name: String? {
    didSet {
      let effectiveName = name ?? ""
      view.text = effectiveName
      guard !effectiveName.isEmpty else { return }
      let task = controller.urlSession.dataTask(
        with: buildSearchRequest(forPlaceName: effectiveName)
      ) {
        data, response, error in
        if let error {
          self.textSearchFailed(with: error)
          return
        }
        do {
          guard let data else {
            preconditionFailure("Response \(response) is empty")
          }
          let decoder = JSONDecoder()
          let decodedResponse = try decoder.decode(PlaceTextSearchResponse.self, from: data)
          if let extractedID = decodedResponse.places.first?.id {
            self.placeID = extractedID
          } else {
            preconditionFailure("Place ID not found from \(response)")
          }
        } catch {
          self.textSearchFailed(with: error)
        }
      }
      task.resume()
    }
  }

  var color: UIColor = .clear {
    didSet {
      colorSelectionButton.setTitleColor(color, for: .normal)
      controller.reloadStyle()
    }
  }

  private(set) var placeID: String? {
    didSet {
      colorSelectionButton.setTitle("⬤", for: .normal)
      controller.reloadStyle()
    }
  }

  private lazy var colorSelectionButton: UIButton = {
    let button = UIButton()
    button.titleLabel?.font = button.titleLabel?.font.withSize(12)
    button.setTitle("❓", for: .normal)
    button.addTarget(self, action: #selector(selectionButtonTapped), for: .touchUpInside)
    return button
  }()

  lazy var view: UITextField = {
    let textField = UITextField()
    textField.translatesAutoresizingMaskIntoConstraints = false
    textField.backgroundColor = .white.withAlphaComponent(0.5)
    textField.font = textField.font?.withSize(12)
    textField.delegate = self

    textField.leftView = colorSelectionButton
    textField.leftViewMode = .always

    let clearButton = UIButton()
    clearButton.titleLabel?.font = clearButton.titleLabel?.font.withSize(12)
    clearButton.setTitle("🗑️", for: .normal)
    clearButton.addTarget(self, action: #selector(remove), for: .touchUpInside)
    textField.rightView = clearButton
    textField.rightViewMode = .unlessEditing

    textField.clearButtonMode = .whileEditing
    return textField
  }()

  init(controller: DataDrivenStylingSearchViewController, serial: Int) {
    self.controller = controller
    self.serial = serial
  }

  func focusTextEdit() {
    view.becomeFirstResponder()
  }

  @objc func remove() {
    controller.removeFeature(self)
  }

  private func textSearchFailed(with error: Error) {
    colorSelectionButton.setTitle("❗", for: .normal)
    NSLog("Feature \(name) not found: \(error)")
  }

  @objc func selectionButtonTapped() {
    let selector = UIColorPickerViewController()
    selector.selectedColor = color ?? .white
    selector.supportsAlpha = false
    selector.delegate = self
    controller.present(selector, animated: true)
  }
}

extension PlaceLookupController: UITextFieldDelegate {
  func textFieldDidEndEditing(_ textField: UITextField) {
    let text = textField.text?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    guard !text.isEmpty else {
      colorSelectionButton.setTitle("❓", for: .normal)
      return
    }
    guard name != text else { return }
    colorSelectionButton.setTitle("⏳", for: .normal)
    name = text
  }

  func textFieldShouldReturn(_ textField: UITextField) -> Bool {
    textField.resignFirstResponder()
    return false
  }
}

extension PlaceLookupController: UIColorPickerViewControllerDelegate {
  func colorPickerViewControllerDidSelectColor(_ viewController: UIColorPickerViewController) {
    color = viewController.selectedColor
  }
}

/// Table view that does not intercept gestures outside table cells.
class NotCapturingTouchesTableView: UITableView {
  // This override causes the view to not intercept gestures, unless the gesture occurs on one of
  // this view's subviews. The UITableView should not prevent the map from panning/zooming/tilting.
  override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
    let view = super.hitTest(point, with: event)
    // If no descendent of this view contains the specified point, return nil. nil is returned if
    // the point does not fit inside the view.
    if view == self {
      return nil
    } else {
      // If there is a descendent of this view that contains the specified point, return that view.
      return view
    }
  }
}

class DataDrivenStylingSearchViewController: UIViewController {
  private let cellIdentifier = "cellIdentifier"

  fileprivate let urlSession = URLSession(
    configuration: .default, delegate: nil,
    delegateQueue: .main)

  private var placeSerial = 0
  private var places: [Int: PlaceLookupController] = [:]
  private var computedColorMapping: [String: UIColor]?

  private lazy var tableView: UITableView = {
    let view = NotCapturingTouchesTableView()
    view.translatesAutoresizingMaskIntoConstraints = false
    view.backgroundView = nil
    view.backgroundColor = .clear
    view.contentInsetAdjustmentBehavior = .never
    view.rowHeight = 28
    view.register(UITableViewCell.self, forCellReuseIdentifier: cellIdentifier)
    return view
  }()

  private lazy var dataSource = UITableViewDiffableDataSource<Int, Int>(tableView: tableView) {
    (tableView: UITableView, indexPath: IndexPath, itemIdentifier: Int) -> UITableViewCell? in
    let cell = tableView.dequeueReusableCell(withIdentifier: self.cellIdentifier, for: indexPath)
    cell.contentView.subviews.forEach { $0.removeFromSuperview() }
    cell.backgroundColor = .clear
    guard let view = self.places[itemIdentifier]?.view else { return cell }
    cell.contentView.addSubview(view)
    NSLayoutConstraint.activate(
      [
        cell.contentView.topAnchor.constraint(equalTo: view.topAnchor),
        cell.contentView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
        cell.contentView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
        cell.contentView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
      ])
    return cell
  }

  private var mapID = mapIDWithAdministrativeAreaLevel2Layer

  private lazy var mapView = {
    let camera = GMSCameraPosition(latitude: 40, longitude: -117.5, zoom: 5.5)
    let view = GMSMapView(
      frame: .zero, mapID: GMSMapID(identifier: mapID), camera: camera)
    view.translatesAutoresizingMaskIntoConstraints = false
    return view
  }()

  override func viewDidLoad() {
    super.viewDidLoad()
    if mapID.isEmpty {
      promptForMapID(description: "with all data-driven styling layers enabled") {
        self.mapID = $0
        self.setUp()
      }
    } else {
      setUp()
    }
  }

  private func setUp() {
    guard mapID.count != 0 else {
      let label = UILabel()
      label.text = "A Map ID is required"
      label.textAlignment = .center
      view = label
      return
    }

    tableView.dataSource = dataSource
    view.addSubview(mapView)
    view.addSubview(tableView)

    NSLayoutConstraint.activate(
      [
        mapView.topAnchor.constraint(equalTo: view.topAnchor),
        mapView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
        mapView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
        mapView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
        tableView.topAnchor.constraint(equalTo: view.topAnchor),
        tableView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
        tableView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
        tableView.heightAnchor.constraint(equalTo: view.heightAnchor, multiplier: 0.5),
      ])

    let addAction = UIAction(
      title: "+"
    ) { _ in self.addButtonTapped() }
    navigationItem.rightBarButtonItem = UIBarButtonItem(primaryAction: addAction)

    var diff = NSDiffableDataSourceSnapshot<Int, Int>()
    diff.appendSections([0])
    diff.appendItems(
      initialSearches.map { addFeature(name: $0).0 }
    )
    dataSource.apply(diff)
  }

  private func addButtonTapped() {
    let (itemIdentifier, lookup) = addFeature()
    var diff = dataSource.snapshot()
    diff.appendItems([itemIdentifier])
    dataSource.apply(diff)
    DispatchQueue.main.async {
      lookup.focusTextEdit()
    }
  }

  private func nextColorHue() -> Float {
    var hueValues: [Float] = places.compactMap { (_, config) in
      var hue = CGFloat.nan
      if config.color.getHue(&hue, saturation: nil, brightness: nil, alpha: nil) {
        if hue >= 0 && hue <= 1 {
          return Float(hue)
        }
      }
      return nil
    }
    var candidateHue = (hueValues.first ?? 0) - 0.5
    if hueValues.count > 1 {
      hueValues.sort()
      let diffs =
        [1 + hueValues.first! - hueValues.last!] + zip(hueValues.dropFirst(), hueValues).map(-)
      let (largestGap, endOfLargestGap) = (zip(diffs, hueValues).max { $0.0 < $1.0 })!
      candidateHue = endOfLargestGap - largestGap / 2
    }
    if candidateHue < 0 {
      candidateHue += 1
    }
    return candidateHue
  }

  private func addFeature(name: String? = nil) -> (Int, PlaceLookupController) {
    placeSerial += 1
    let itemIdentifier = placeSerial
    let featureConfig = PlaceLookupController(controller: self, serial: itemIdentifier)
    featureConfig.color = UIColor(
      hue: CGFloat(nextColorHue()),
      saturation: 1,
      brightness: 0.75,
      alpha: 1)
    featureConfig.name = name
    places[itemIdentifier] = featureConfig
    return (itemIdentifier, featureConfig)
  }

  fileprivate func removeFeature(_ feature: PlaceLookupController) {
    guard places.count > 1 else { return }
    var diff = dataSource.snapshot()
    diff.deleteItems([feature.serial])
    dataSource.apply(diff)

    places.removeValue(forKey: feature.serial)
    reloadStyle()
  }

  fileprivate func reloadStyle() {
    var colorMapping: [String: UIColor] = Dictionary(
      places.compactMap { (_, config) in
        config.placeID.map { ($0, config.color) }
      },
      uniquingKeysWith: { (first, _) in first })
    // Skip restyling if the mapping hasn't actually changed, since restyling can be expensive.
    guard colorMapping != computedColorMapping else { return }
    computedColorMapping = colorMapping
    mapView.featureLayer(of: .administrativeAreaLevel2).style = {
      guard let color = colorMapping[$0.placeID] else { return nil }
      return FeatureStyle(
        fill: color.withAlphaComponent(0.5),
        stroke: color,
        strokeWidth: 1.5
      )
    }
  }
}
